### PR TITLE
Docker support

### PR DESCRIPTION
**Pull request by sandhawalia**
This PR adresses https://github.com/moabitcoin/cherry-pytorch/issues/37

### CODE REVIEW

File: Dockerfile.cpu [hunk 1]
```diff
@@ -0,0 +1,62 @@
+FROM ubuntu:18.04 as build-stage
+
+WORKDIR /usr/src/app
+
+ENV LANG="C.UTF-8" LC_ALL="C.UTF-8" PATH="/opt/venv/bin:$PATH" PIP_NO_CACHE_DIR="false" CFLAGS="-mavx2 -mf16c" CXXFLAGS="-mavx2 -mf16c"
```
Comment (daniel-j-h) [inline Dockerfile.cpu]: Don't think you need the cflags / cxxflags since you don't compile anything from source

File: Dockerfile.cpu [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+
+RUN apt-get update && DEBIAN_FRONTEND=noninteractive apt-get install -y --no-install-recommends \
+    python3 python3-pip python3-venv python3-dev wget make g++ libblas-dev liblapack-dev patch \
+    build-essential zlib1g-dev libsdl2-dev libjpeg-dev nasm tar libbz2-dev libgtk2.0-dev cmake \
+    git libfluidsynth-dev libgme-dev libopenal-dev timidity libwildmidi-dev unzip libboost-all-dev \
+    libcanberra-gtk-module libopenmpi-dev libosmesa6-dev libgl1-mesa-glx libglfw3 && \
+    rm -rf /var/lib/apt/lists/*
```
Comment (daniel-j-h) [inline Dockerfile.cpu]: Not sure why you need all these -dev packages, and make, gcc, etc.? What do you compile from source? Is it python dependencies compiling from source?

File: Dockerfile.cpu [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+
+RUN export DISPLAY=:0
```
Comment (daniel-j-h) [inline Dockerfile.cpu]: You can fold this line into the ENV command above

File: Dockerfile.cpu [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+WORKDIR /usr/src/app
+
+ENV LANG="C.UTF-8" LC_ALL="C.UTF-8" PATH="/opt/venv/bin:$PATH" PIP_NO_CACHE_DIR="false"
+
+RUN apt-get update && DEBIAN_FRONTEND=noninteractive apt-get install -y --no-install-recommends \
+    python3 python3-pip python3-venv python3-dev wget make g++ libblas-dev liblapack-dev patch \
+    build-essential zlib1g-dev libsdl2-dev libjpeg-dev nasm tar libbz2-dev libgtk2.0-dev cmake \
+    git libfluidsynth-dev libgme-dev libopenal-dev timidity libwildmidi-dev unzip libboost-all-dev \
+    libcanberra-gtk-module libopenmpi-dev libosmesa6-dev libgl1-mesa-glx libglfw3 && \
+    rm -rf /var/lib/apt/lists/*
```
Comment (daniel-j-h) [inline Dockerfile.cpu:12]: If you do compile from source above, then here you don't need the -dev packages anymore.

For package p
- libp-dev contains headers, impl., sources, etc. for compilation and linking
- libp contains runtime libs for dynamic loading

Use the dev packages for your compilation stage and the non-dev packages for your final stage.

File: Dockerfile.cpu [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+
+RUN export DISPLAY=:0
```
Comment (daniel-j-h) [inline Dockerfile.cpu]: Save as above: use single ENV command

File: Dockerfile.cpu [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+          https://download.pytorch.org/whl/cpu/torchvision-0.4.1%2Bcpu-cp36-cp36m-linux_x86_64.whl  \
+          --hash=sha256:593ad12c3c8ce16068566c9eb2bfb39f4834c89a2cc9f0b181e9121b06046b3e            \
+          \n" >> requirements.txt
+
+RUN python3 -m piptools sync
+
+RUN git clone https://github.com/openai/baselines.git && \
+    cd baselines && \
+    python3 -m pip install -e . && \
+    rm -rf baselines
```
Comment (daniel-j-h) [inline Dockerfile.cpu]: Capturing from chat:
- get tarball snapshot
- verify checksum

and always (!) pin deps to specific versions, gitshas, and related.

File: Dockerfile.gpu [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+
+RUN git clone https://github.com/openai/baselines.git && \
+    cd baselines && \
+    pip install -e . && \
+    rm -rf baselines
+
+COPY . .
+
+ENTRYPOINT ["/usr/src/app/bin/cherry"]
+CMD ["-h"]
```
Comment (daniel-j-h) [inline Dockerfile.gpu]: Looks like this is mostly a copy of the CPU Dockerfile with a different FROM image. Not adding comments here; the CPU comments apply here, too.

File: README.md [hunk 8]
```diff
@@ -1,8 +1,11 @@
...
+Reinforcement Learning &amp; other bedtime stories in PyTorch. This repository is [pytorch](https://pytorch.org) implementation of the [Intro RL course](https://github.com/simoninithomas/Deep_reinforcement_learning_Course) & [Advanced RL course](https://github.com/dennybritz/reinforcement-learning). Before training/playing the agent please install the external [dependencies.](https://github.com/moabitcoin/cherry-pytorch#dependencies) The name cherry is derived from Yann Lecun's [statement](https://www.quora.com/What-do-you-think-about-reinforcement-learning-Is-it-the-cherry-on-the-cake-as-Yann-LeCun-puts-it-1) regarding RL being the cherry on the ML [cake.](https://miro.medium.com/max/1200/1*bvMhd_xpVxfJYoKXYp5hug.png).
 
 ## :computer: Installation
+```
+Tested with ubuntu:18.04 only
```
Comment (daniel-j-h) [inline README.md:15]: Docker instructions missing; maybe change all the conda and manual installation instructions to just using docker?

File: bin/cherry [hunk 9]
```diff
@@ -1,4 +1,4 @@
-#!/usr/bin/env python3
+#!/bin/bash
```
Comment (daniel-j-h) [inline bin/cherry]: Don't think this is correct. Below you are using Python syntax to import a module; so this file has to be executed by the Python interpreter, no?

Or was the idea to switch to bash but then do something like

    python3 python3 -m cherry.cli "$@"

?

https://github.com/moabitcoin/ig65m-pytorch/blob/fc749e2ee354c3e4ddbb144cf511bb868b008f61/bin/ig65m

File: requirements.in [hunk 10]
```diff
@@ -1,9 +1,6 @@
...
-pip-tools
+GitPython
 pyyaml
-argparse
+scikit-image
 scikit-video
-ipython
-pyglet
-GitPython
+tensorflow_gpu
```
Comment (daniel-j-h) [inline requirements.in:5]: the :cake: is a lie!

File: Dockerfile.cpu (hunk 1 above)
Comment (sandhawalia) [reply Dockerfile.cpu]: We do. [Vizdoom](https://github.com/mwydmuch/ViZDoom/blob/master/doc/Building.md#linux_deps)

File: Dockerfile.cpu (hunk 2 above)
Comment (sandhawalia) [reply Dockerfile.cpu]: Blame [Vizdoom](https://github.com/mwydmuch/ViZDoom/blob/master/doc/Building.md#linux_deps) again

File: Dockerfile.cpu (hunk 3 above)
Comment (sandhawalia) [reply Dockerfile.cpu]: Indeed !

File: Dockerfile.cpu (hunk 4 above)
Comment (sandhawalia) [reply Dockerfile.cpu:12]: Thats current. Replicated Vizdoom's interactions here. Will modify

File: Dockerfile.cpu (hunk 5 above)
Comment (sandhawalia) [reply Dockerfile.cpu]: Done

File: README.md (hunk 8 above)
Comment (sandhawalia) [reply README.md:15]: Wanna keep both. :)

File: Dockerfile.cpu (hunk 6 above)
Comment (sandhawalia) [reply Dockerfile.cpu]: Changed

File: Dockerfile.cpu (hunk 5 above)
Comment (sandhawalia) [reply Dockerfile.cpu]: Done

File: requirements.in (hunk 10 above)
Comment (sandhawalia) [reply requirements.in:5]: There is no ☕️ without the 🍰

File: Dockerfile.cpu [hunk 11]
```diff
@@ -0,0 +1,61 @@
+FROM ubuntu:18.04 as build-stage
+
+WORKDIR /usr/src/app
+
+ENV LANG="C.UTF-8" LC_ALL="C.UTF-8" PATH="/opt/venv/bin:$PATH" PIP_NO_CACHE_DIR="false" DISPLAY=":0" CXXFLAGS="-I/opt/venv/lib/python3.6/site-packages/numpy/core/include"
```
Comment (daniel-j-h) [inline Dockerfile.cpu]: This sets the C++ include dir only; not sure it's what you want.

Maybe better to set / append to global env vars

    C_INCLUDE_DIR
    CXX_INCLUDE_DIR

and then also LD_LIBRARY_DIR

(if I remember correctly from C/CXX land)

File: Dockerfile.cpu [hunk 12]
```diff
@@ -0,0 +1,61 @@
...
+COPY requirements.txt .
+
+RUN python3 -m venv /opt/venv && \
+    python3 -m pip install pip==19.3.1 pip-tools==4.2.0 wheel==0.33.6 setuptools==44.0.0
+
+RUN python3 -m piptools sync
+
+COPY . .
+
+FROM ubuntu:18.04 as dev-stage
```
Comment (daniel-j-h) [inline Dockerfile.cpu:1]: I don't understand the use of two stages here. Your second stage never uses artifacts produced in the first stage?

That is, I'm missing a `COPY --from=build-stage` below.

The idea with multi-stage builds usually is to compile and link everything you need in the build stage, then e.g. build a self-contained wheel. And in your last stage (which no longer contains compilers, linkers, etc.) you copy this wheel over from the build stage.

File: Dockerfile.cpu [hunk 13]
```diff
@@ -0,0 +1,61 @@
...
+    unzip baselines.zip && \
+    cd baselines*/ && \
+    python3 -m pip install . && \
+    cd ../ && \
+    rm -rf baselines*
+
+COPY . .
+
+RUN python3 -m venv /opt/venv && \
+    python3 -m pip install Pillow==6.2.1 && \
```
Comment (daniel-j-h) [inline Dockerfile.cpu:38]: Why installing Pillow here and not in the requirements? Also would Pillow-SIMD and/or libjpeg-turbo be of any use here to get some :fire: behind the Pillow :rocket: ?

File: README.md [hunk 14]
```diff
@@ -1,20 +1,32 @@
...
+# After installed external dependencies
 conda create -n cherry python=3.6
 conda activate cherry
 pip install -r requirements.txt
 pip install -e .
 ```
 ### :whale: Docker
+```
+# CPU
+make install dockerfile=Dockerfile.cpu dockerimage=moabitcoin/cherry-pytorch-cpu
```
Comment (daniel-j-h) [inline README.md:28]: Can this be simplified to

    make

since the rest is set by default already, no?

File: README.md [hunk 15]
```diff
@@ -1,20 +1,32 @@
...
+make run dockerimage=moabitcoin/cherry-pytorch-cpu
```
Comment (daniel-j-h) [inline README.md:29]: Same here, `make run` by default should do the Right Thing :tm:

Comment (daniel-j-h) [reply README.md:29]: Only for GPU version it will be different.

File: Dockerfile.cpu (hunk 13 above)
Comment (sandhawalia) [reply Dockerfile.cpu:38]: Pillow defaults to 7.0.0 from requirements.in/txt which is incompatible with Torchvision 🙈

File: README.md (hunk 14 above)
Comment (sandhawalia) [reply README.md:28]: just to easily clarify diff in cpu/gpu with input args.

File: Dockerfile.cpu (hunk 12 above)
Comment (sandhawalia) [reply Dockerfile.cpu:1]: yes this can go :)

File: Dockerfile.cpu (hunk 11 above)
Comment (sandhawalia) [reply Dockerfile.cpu]: Nice  ! Trying this out now 👍

Comment (sandhawalia) [reply Dockerfile.cpu]: > This sets the C++ include dir only;
Yes just need the numpy headers

File: Dockerfile.cpu (hunk 13 above)
Comment (daniel-j-h) [reply Dockerfile.cpu:38]: Ah, okay. Then the best way to express this would be in requirements.in something like

    Pillow < 7

**Pull request merged**